docs: add guide for CI/CD template testing

matifali opened the pull request:
Introduce a new tutorial outlining the process for testing and publishing
Coder templates within a CI/CD pipeline using GitHub Actions. This guide
includes prerequisites, a comprehensive workflow example, and detailed
steps for template validation and deployment.

Closes #9899

matifali committed:
docs: add guide for CI/CD template testing

Introduce a new tutorial outlining the process for testing and publishing
Coder templates within a CI/CD pipeline using GitHub Actions. This guide
includes prerequisites, a comprehensive workflow example, and detailed
steps for template validation and deployment.

matifali committed:
Update manifest to include testing templates guide

Add a new entry in the documentation's manifest for a tutorial on testing
and publishing Coder templates in a CI/CD pipeline.

bpmct commented on docs/tutorials/testing-templates.md:
```diff
@@ -0,0 +1,104 @@
+        run: |
+          coder templates push $TEMPLATE_NAME --activate=false --name ${{ steps.name.outputs.version_name }} --message "${{ steps.message.outputs.pr_title }}" --yes
+
+      - name: Create a test workspace
+        run: |
+          coder create -t $TEMPLATE_NAME --template-version ${{ steps.name.outputs.version_name }} --yes
```
Can we also show how to test if specific conditions work inside the template (e.g. `make build`) or something? Like how to run example commands and ensure they run properly? The `coder exp scaletest` command may be useful for this.

bpmct reviewed (commented):
Would love @jatcod3r's review as this was something we were discussing lately. We could add this to our demo env

matifali committed:
Enhance template testing with SSH command execution

Add SSH configuration and example commands to test the workspace creation using the new template version. This ensures the template functions as expected before promotion.

bpmct commented:
@jatcod3r Can you be the approver for this PR?

EdwardAngert commented on docs/admin/templates/managing-templates/change-management.md:
````diff
@@ -88,6 +88,14 @@ coder templates push --yes $CODER_TEMPLATE_NAME \
     --name=$CODER_TEMPLATE_VERSION # Version name is optional
 ```
 
+You can also use the [coder/setup-action](https://github.com/coder/setup-coder)
+GitHub Action to install the Coder CLI and push new template versions.
````
we use `You can also` at the beginning and end of this section. If this is an alternative to the steps above, can we move this sentence to the top of the section so that users can decide which path to take before we show them the steps?

EdwardAngert commented on docs/manifest.json:
```diff
@@ -709,6 +709,11 @@
 					"path": "./tutorials/cloning-git-repositories.md"
 				},
+				{
+					"title": "Testing Templates",
```
```suggestion
					"title": "Test Templates Through CI/CD",
```

EdwardAngert commented on docs/tutorials/testing-templates.md:
```diff
@@ -0,0 +1,110 @@
+# Testing and Publishing Coder Templates in CI/CD
```
```suggestion
# Test and Publish Coder Templates Through CI/CD
```

EdwardAngert commented on docs/tutorials/testing-templates.md:
```diff
@@ -0,0 +1,110 @@
+
+## Prerequisites
+
+Before proceeding, ensure the following:
+
+- **Coder CLI** is installed and configured in your environment.
+- **Terraform CLI** is installed and available in your CI environment.
+- Access to your **Coder instance** with the appropriate
+  [permissions](../admin/users/groups-roles.md#roles).
+
```
```suggestion
- Install and configure Coder CLI in your environment.
- Install Terraform CLI in your CI environment.
- Configure [user roles and permissions](../admin/users/groups-roles.md#roles) for your Coder instance.

```

EdwardAngert commented on docs/tutorials/testing-templates.md:
```diff
@@ -0,0 +1,110 @@
+
+## Example GitHub Action Workflow
+
+Below is an example workflow for testing and publishing a template using GitHub
+Actions. The workflow first validates the Terraform template, pushes the
+template to Coder without activating it, tests the template by creating a
+workspace, and then promotes the template version to active upon successful
+workspace creation.
+
+### Step-by-Step Process
+
+1. **Validate the Terraform template.**
+2. **Push the template to Coder without activating it.**
+3. **Test the template by creating a workspace.**
+4. **Promote the template version to active upon successful workspace
+   creation.**
```
```suggestion
This example workflow tests and publishes a template using GitHub
Actions.

The workflow:

1. Validates the Terraform template.
1. Pushes the template to Coder without activating it.
1. Tests the template by creating a workspace.
1. Promotes the template version to active upon successful workspace creation.
```

matifali committed:
Apply suggestions from code review

Co-authored-by: Edward Angert <EdwardAngert@users.noreply.github.com>

matifali committed:
Merge branch 'main' into atif/test-templates

bpmct commented on docs/tutorials/testing-templates.md:
```diff
@@ -0,0 +1,103 @@
+        run: |
+          coder templates push $TEMPLATE_NAME --activate=false --name ${{ steps.name.outputs.version_name }} --message "${{ steps.message.outputs.pr_title }}" --yes
+
+      - name: Create a test workspace
+        run: |
+          coder create -t $TEMPLATE_NAME --template-version ${{ steps.name.outputs.version_name }} test-${{ steps.name.outputs.version_name }} --yes
```
Should we delete the test workspace?

matifali replied:
Yes. Good catch. I will do.

matifali committed:
Update template testing workflow to ensure cleanup

matifali committed:
`make fmt`

matifali committed:
docs: update CLI instructions and link setup-coder

Clarified CLI installation instructions for pushing template versions
and linked to the GitHub Actions setup-coder for automation. Removed
redundant setup-action note and added details on creating a headless
user for CI workflows.

EdwardAngert reviewed (approved):
🚀

jatcod3r commented on docs/admin/templates/managing-templates/change-management.md:
Could we write somewhere something about creating a user that only has template push and workspace create/delete permissions? I don't think an admin would solely rely on their own token to just push + test templates, it would typically be handled by an account with least privilege. 

The current workflow I follow is creating a password-less user that can only push templates, create and delete a workspace. I then create a token for it over the REST API (Coder CLI doesn't work for this case) and then follow the rest of this doc as is.

matifali replied:
Coder CLi supports creating a token for another user.

`coder token create --user machine --lifetime 8760h`

I am going to add this.

jatcod3r replied:
yea sorry about that 😅 I was running it in v2.16 when the --user flag would return `parsing flags ([tokens list --user=headless]) for "coder tokens create": unknown flag: --user`. That works now!

jatcod3r commented on docs/admin/templates/managing-templates/change-management.md:
Looks good now!
